Add coordinate system conversion in autoware_demo.py

Goldob opened the pull request:
Before the change, when Autoware initializes localization when running with CARLA, we can see the following issues:
- the initial pose is facing the wrong direction;
- the initial pose is slightly tilted with respect to the road axis.

<img width="1920" height="1080" alt="Screenshot_2025-09-04_00-36-29" src="https://github.com/user-attachments/assets/4a5d5b05-cdb3-4f25-9a17-7e83e74f7f0a" />

This has been fixed by inverting the yaw angles of the following transforms when spawning the vehicle in CARLA:
- `velodyne_top` to `sensor_kit` (fixes the first issue)
- `sensor_kit_base_link` to `base_link` (fixes the second issue)

After the change is applied, we can see the Autoware localizing correctly:

<img width="1920" height="1080" alt="Screenshot_2025-09-04_00-45-00" src="https://github.com/user-attachments/assets/2aa8849f-3195-464a-81da-18fd992c4a07" />

As described in #9, current sensor positions are taken directly from URDF. The discrepancy might be due to different angle conventions between Autoware and CARLA, but it has not been investigated further. Convention for other angles (roll and pitch) were also not checked.

Goldob committed:
Invert yaw angles used in autoware_demo.py

TauTheLepton commented:
Carla uses different frame of reference than ROS - mainly the Y axis is pointing to the other side (e.g. PCD has to be adjusted by inverting all points Y values) - this may be the cause of the described issues.
If spawning values are taken directly from URDF then they are in the ROS-style frame, and they are passed to CARLA which expects Unreal-style frame.

TauTheLepton committed:
Revert "Invert yaw angles used in autoware_demo.py"

This reverts commit 692c10844dc1dbfe021041164ee7e2bcb1ef2f55.

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton committed:
Add coordinate systems conversion class for ROS -> CARLA

This allows to define everything in script using ROS system and have it automatically converted

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton commented:
I have proposed a conversion class to define all transforms the same as in URDF (in ROS system) and convert to Unreal system.
I am not 100% sure the rotation conversions are correct, but they seem to work.

> [!NOTE]
> This does not solve the issue with Autoware not knowing that it is turning. Thit issue is caused by bad IMU readings.

TauTheLepton committed:
Fix rotation conversion and add source to note

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton committed:
Add ros_name to intermediate actors in URDF-like structure of Ego vehicle

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton commented:
I have updated the transform conversion to account for different Z-axis rotation direction (than X and Y-axis) in UnrealUngine

TauTheLepton committed:
Merge remote-tracking branch 'robotec/autoware-support' into fix/frame-rotations

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton committed:
Add manual transformations to attach IMU sensor directly to Ego vehicle

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton committed:
Use new ROS2.Transform class

Signed-off-by: Mateusz Palczuk <mateusz.palczuk@robotec.ai>

TauTheLepton commented:
I added transformation calculations to the demo script to attach IMU directly to Ego vehicle, as this is required for thee IMU to work correctly (for angular velocities).

After this change and #28 the result is that Autoware turns in good direction and recognizes the turn, but does not quite arrive at the goal (see the picture below - this is after a 90-degree left turn).

<img width="835" height="909" alt="image" src="https://github.com/user-attachments/assets/655d0101-1579-49d6-b3c9-76bd5a3daefe" />

This may be caused by wrong vehicle model being used in Autoware. We could probably get better results by setting vehicle settings in vehicle model according to the vehicle actually used. I guess wrong turning may be caused by wrong wheelbase and other vehicle parameters being different in Carla and what Autoware thinks.

### Video

https://github.com/user-attachments/assets/87c727b5-b98f-45ba-979d-70ea25e2b184

_NOTE: Video is recorded with time scale of 0.25_
_NOTE 2: Video contains the fix from #27 - the camera is normal brightness_
